definitions/npm/react-dnd_v2.x.x/flow_>=v0.23.x/react-dnd_v2.x.x.js

```diff
@@ -1,6 +1,202 @@
+// Shared
+// ----------------------------------------------------------------------
+// todo: add symbol once flow supports it
+type Identifier = string;
+
+type ClientOffset = {
+  x: number,
+  y: number
+}
+
+type DndOptions<P> = {
+  arePropsEqual?: (props: P, otherProps: P) => boolean
+};
+
+// Decorated Components
+// ----------------------------------------------------------------------
+declare class DndComponent<P> extends React$Component<any, P, any> {
+  static DecoratedComponent: Class<this>;
+  getDecoratedComponentInstance(): this;
+  getHandlerId(): Identifier;
+}
+
+declare class ContextComponent<P> extends React$Component<any, P, any> {
+  static DecoratedComponent: Class<this>;
+  getDecoratedComponentInstance(): this;
+  // getManager is not yet documented in ReactDnd Docs
+  getManager(): any;
+}
+
+// Drag Source
+// ----------------------------------------------------------------------
+type DragSourceType<P> =
+  Identifier |
+  (props: P) => Identifier;
+
+type DragSourceSpec<P> = {
+  beginDrag: (
+    props: P,
+    monitor?: DragSourceMonitor,
+    component?: React$Component<any, P, any>
+  ) => Object,
+
+  endDrag?: (
+    props: P,
+    monitor?: DragSourceMonitor,
+    component?: ?React$Component<any, P, any>
+  ) => void,
+
+  canDrag?: (
+    props: P,
+    monitor?: DragSourceMonitor
+  ) => boolean,
+
+  isDragging?: (
+    props: P,
+    monitor?: DragSourceMonitor
+  ) => boolean
+};
+
+type DragSourceMonitor = {
+  canDrag: () => boolean,
+  isDragging: () => boolean,
+  getItemType: () => Identifier,
+  getItem: () => Object,
+  getDropResult: () => Object,
+  didDrop: () => boolean,
+  getInitialClientOffset: () => ClientOffset,
+  getInitialSourceClientOffset: () => ClientOffset,
+  getClientOffset: () => ClientOffset,
+  getDifferenceFromInitialOffset: () => ClientOffset,
+  getSourceClientOffset: () => ClientOffset
+}
+
+type DragSourceConnector = {
+  dragSource: () => ConnectDragSource,
+  dragPreview: () => ConnectDragPreview
+}
+
+type DragSourceOptions = {
+  dropEffect?: string
+}
+
+type DragPreviewOptions = {
+  captureDraggingState?: boolean,
+  anchorX?: number,
+  anchorY?: number
+}
+
+type ConnectDragSource = (
+  elementOrNode: React$Element<any>
+  options?: DragSourceOptions
+) => React$Element<any>;
+
+type ConnectDragPreview = (
+  elementOrNode: React$Element<any>
+  options?: DragPreviewOptions
+) => React$Element<any>;
+
+type DragSourceCollector = (
+  connect: DragSourceConnector,
+  monitor: DragSourceMonitor
+) => Object;
+
+type DragSource = <P, C: React$Component<any, P, any>>(
+  type: DragSourceType<P>,
+  spec: DragSourceSpec<P>,
+  collect: DragSourceCollector,
+  options?: DndOptions<P>
+) => (component: Class<C>) => Class<DndComponent<P>>;
```

Comment:
I'm not completely sure about this one either. I added a test in which DragSource is applied to a functional React components and things seem to still type check.